// enka-network-api/dist/models/artifact/ArtifactData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArtifactData = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const ImageAssets_1 = require("../assets/ImageAssets");
const TextAssets_1 = require("../assets/TextAssets");
const ArtifactSet_1 = require("./ArtifactSet");
class ArtifactData {
    constructor(data, enka, setData) {
        var _a;
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("descTextMapHash"), enka);
        this.equipType = json.getAsString("equipType");
        const equipTypeJson = (_a = enka.cachedAssetsManager.getGenshinCacheData("ManualTextMapConfigData").findArray((_, p) => p.getAsString("textMapId") === this.equipType)) === null || _a === void 0 ? void 0 : _a[1];
        if (!equipTypeJson)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Artifact Equip Type", this.equipType);
        this._equipTypeData = equipTypeJson.getAsJsonObject();
        this.equipTypeName = new TextAssets_1.TextAssets(equipTypeJson.getAsNumber("textMapContentTextMapHash"), enka);
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("icon"), enka);
        this.stars = json.getAsNumber("rankLevel");
        this.set = setData ? new ArtifactSet_1.ArtifactSet(setData, enka) : ArtifactSet_1.ArtifactSet.getById(json.getAsNumber("setId"), enka);
    }
    static getById(id, enka, setData) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("ReliquaryExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Artifact", id);
        return new ArtifactData(json.getAsJsonObject(), enka, setData);
    }
}
exports.ArtifactData = ArtifactData;
