// enka-network-api/dist/models/theater/TheaterMode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TheaterMode = exports.theaterDifficulties = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
exports.theaterDifficulties = ["EASY", "NORMAL", "HARD", "VISIONARY"];
class TheaterMode {
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("difficultyId");
        this.difficulty = exports.theaterDifficulties[json.getAsNumber("difficultyLevel") - 1];
    }
    static getById(id, enka) {
        var _a;
        const theaterMode = (_a = enka.cachedAssetsManager.getGenshinCacheData("RoleCombatDifficultyExcelConfigData").findArray((_, p) => p.getAsNumber("difficultyId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!theaterMode)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("TheaterMode", id);
        return new TheaterMode(theaterMode.getAsJsonObject(), enka);
    }
}
exports.TheaterMode = TheaterMode;
