// enka-network-api/dist/client/EnkaClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnkaClient = exports.defaultEnkaClientOptions = exports.defaultImageBaseUrls = void 0;
const GenshinUser_1 = require("../models/GenshinUser");
const characterUtils = __importStar(require("../utils/character_utils"));
const CachedAssetsManager_1 = require("./CachedAssetsManager");
const CharacterData_1 = require("../models/character/CharacterData");
const WeaponData_1 = require("../models/weapon/WeaponData");
const Costume_1 = require("../models/character/Costume");
const axios_utils_1 = require("../utils/axios_utils");
const Material_1 = require("../models/material/Material");
const ArtifactData_1 = require("../models/artifact/ArtifactData");
const DetailedGenshinUser_1 = require("../models/DetailedGenshinUser");
const enka_system_1 = require("enka-system");
const GenshinCharacterBuild_1 = require("../models/enka/GenshinCharacterBuild");
const Material_2 = require("../models/material/Material");
const ArtifactSet_1 = require("../models/artifact/ArtifactSet");
const config_file_js_1 = require("config_file.js");
const getUserUrl = (enkaUrl, uid) => `${enkaUrl}/api/uid/${uid}`;
const userCacheMap = new Map();
exports.defaultImageBaseUrls = [
    {
        url: "https://homdgcat.wiki/homdgcat-res",
        priority: 12,
        format: "PNG",
        regexList: [
            /^UI_(AvatarIcon|RelicIcon|EquipIcon|ItemIcon)_/,
            /^Skill_/,
            /^UI_Gacha_AvatarImg_/,
            /^UI_NameCardPic_(.+)_P$/,
        ],
        customParser(fileName) {
            if (fileName.startsWith("UI_AvatarIcon_")) {
                return `Avatar/${fileName}.png`;
            }
            else if (fileName.startsWith("UI_RelicIcon_")) {
                return `Relic/${fileName}.png`;
            }
            else if (fileName.startsWith("UI_EquipIcon_")) {
                return `Weapon/${fileName}.png`;
            }
            else if (fileName.startsWith("UI_ItemIcon_")) {
                return `Mat/${fileName}.png`;
            }
            else if (fileName.startsWith("Skill_")) {
                return `AvatarSkill/${fileName}.png`;
            }
            else if (fileName.startsWith("UI_Gacha_AvatarImg_")) {
                return `Gacha/${fileName}.png`;
            }
            else if (fileName.startsWith("UI_NameCardPic_") && fileName.endsWith("_P")) {
                return `Avatar/${fileName}.png`;
            }
            else {
                throw new Error(`Unhandled file name: ${fileName}`);
            }
        },
    },
    {
        url: "https://enka.network/ui",
        priority: 10,
        format: "PNG",
        regexList: [
            /^UI_(Costume|NameCardIcon|NameCardPic|RelicIcon|AvatarIcon_Side|EquipIcon)_/,
            /^UI_AvatarIcon_(.+)_(Card|Circle)$/,
        ],
    },
    {
        url: "https://gi.yatta.moe/assets/UI",
        priority: 3,
        format: "PNG",
        regexList: [
            /.*/,
        ],
    },
    {
        url: "https://api.hakush.in/gi/UI",
        priority: 2,
        format: "WEBP",
        regexList: [
            /.*/,
        ],
    },
];
exports.defaultEnkaClientOptions = {
    "enkaUrl": "https://enka.network",
    "imageBaseUrls": [...exports.defaultImageBaseUrls],
    "userAgent": "Mozilla/5.0",
    "requestTimeout": 3000,
    "defaultLanguage": "en",
    "textAssetsDynamicData": {
        paramList: [],
        userInfo: null,
    },
    "cacheDirectory": null,
    "showFetchCacheLog": true,
    "userCache": {
        isEnabled: true,
        getter: null,
        setter: null,
        deleter: null,
    },
    "githubToken": null,
    "enkaSystem": null,
};
class EnkaClient {
    getUser(data) {
        const fixedData = (0, config_file_js_1.renameKeys)(data, { "player_info": "playerInfo" });
        return new GenshinUser_1.GenshinUser(fixedData, this);
    }
    getCharacterBuild(data, username, hash) {
        return new GenshinCharacterBuild_1.GenshinCharacterBuild(data, this, username, hash);
    }
    /** @param options Options for the client */
    constructor(options = {}) {
        this._tasks = [];
        this.hoyoType = 0;
        const mergedOptions = (0, config_file_js_1.bindOptions)(exports.defaultEnkaClientOptions, options);
        if (!mergedOptions.enkaSystem) {
            if (enka_system_1.EnkaSystem.instance.getLibrary(this.hoyoType)) {
                mergedOptions.enkaSystem = new enka_system_1.EnkaSystem();
            }
            else {
                mergedOptions.enkaSystem = enka_system_1.EnkaSystem.instance;
            }
        }
        this.options = mergedOptions;
        const userCacheFuncs = [this.options.userCache.getter, this.options.userCache.setter, this.options.userCache.deleter];
        if (userCacheFuncs.some(f => f) && userCacheFuncs.some(f => !f))
            throw new Error("All user cache functions (setter/getter/deleter) must be null or all must be customized.");
        this.cachedAssetsManager = new CachedAssetsManager_1.CachedAssetsManager(this);
        this.options.enkaSystem.registerLibrary(this);
    }
    /**
     * @param uid In-game UID of the user
     * @param collapse Whether to fetch rough user information (Very fast)
     * @returns DetailedGenshinUser if collapse is false, GenshinUser if collapse is true
     * @throws {EnkaNetworkError}
     */
    async _fetchUser(uid, collapse) {
        var _a, _b, _c, _d;
        if (isNaN(Number(uid)))
            throw new Error("Parameter `uid` must be a number or a string number.");
        const cacheGetter = (_a = this.options.userCache.getter) !== null && _a !== void 0 ? _a : (async (key) => userCacheMap.get(key));
        const cacheSetter = (_b = this.options.userCache.setter) !== null && _b !== void 0 ? _b : (async (key, data) => { userCacheMap.set(key, data); });
        const cacheDeleter = (_c = this.options.userCache.deleter) !== null && _c !== void 0 ? _c : (async (key) => { userCacheMap.delete(key); });
        const cacheKey = `${uid}${collapse ? "-info" : ""}`;
        const cachedUserData = (_d = (collapse ? await cacheGetter(cacheKey) : null)) !== null && _d !== void 0 ? _d : await cacheGetter(uid.toString());
        const useCache = !!(cachedUserData && this.options.userCache.isEnabled);
        let data;
        if (!useCache) {
            const url = getUserUrl(this.options.enkaUrl, uid) + (collapse ? "?info" : "");
            const response = await (0, axios_utils_1.fetchJSON)(url, this, true);
            if (response.status !== 200) {
                switch (response.status) {
                    case 400:
                        throw new enka_system_1.InvalidUidFormatError(Number(uid), response.status, response.statusText);
                    case 424:
                        throw new enka_system_1.EnkaNetworkError("Request to enka.network failed because it is under maintenance.", response.status, response.statusText);
                    case 429:
                        throw new enka_system_1.EnkaNetworkError("Rate Limit reached. You reached enka.network's rate limit. Please try again in a few minutes.", response.status, response.statusText);
                    case 404:
                        throw new enka_system_1.UserNotFoundError(`User with uid ${uid} was not found. Please check whether the uid is correct. If you find the uid is correct, it may be a internal server error.`, response.status, response.statusText);
                    default:
                        throw new enka_system_1.EnkaNetworkError(`Request to enka.network failed with unknown status code ${response.status} - ${response.statusText}\nRequest url: ${url}`, response.status, response.statusText);
                }
            }
            // TODO: use structuredClone
            data = Object.assign({}, response.data);
            if (this.options.userCache.isEnabled) {
                const lifetime = data.ttl * 1000;
                const now = Date.now();
                data._lib = { cache_id: (0, config_file_js_1.generateUuid)(), created_at: now, expires_at: now + lifetime, original_ttl: data.ttl };
                const task = setTimeout(async () => {
                    const dataToDelete = await cacheGetter(cacheKey);
                    if (!dataToDelete)
                        return;
                    if (dataToDelete._lib.cache_id === data._lib.cache_id) {
                        await cacheDeleter(cacheKey);
                    }
                    this._tasks.splice(this._tasks.indexOf(task), 1);
                }, lifetime);
                this._tasks.push(task);
                if (!collapse)
                    await cacheDeleter(`${uid}-info`);
                await cacheSetter(cacheKey, data);
            }
        }
        else {
            // TODO: use structuredClone
            data = Object.assign({}, cachedUserData);
            if (collapse)
                delete data["avatarInfoList"];
            data.ttl = Math.ceil((data._lib.expires_at - Date.now()) / 1000);
        }
        // console.log("useCache", useCache);
        const userData = (0, config_file_js_1.bindOptions)(data, { _lib: { is_cache: useCache } });
        const user = collapse ? new GenshinUser_1.GenshinUser(userData, this) : new DetailedGenshinUser_1.DetailedGenshinUser(userData, this);
        return user;
    }
    /**
     * @param uid In-game UID of the user
     */
    async fetchUser(uid) {
        return await this._fetchUser(uid, false);
    }
    /**
     * @param uid In-game UID of the user
     */
    async fetchCollapsedUser(uid) {
        return await this._fetchUser(uid, true);
    }
    /**
     * @param username enka.network username, not in-game nickname
     * @returns the genshin accounts added to the Enka.Network account
     */
    async fetchEnkaGenshinAccounts(username) {
        return await this.options.enkaSystem.fetchEnkaGameAccounts(username, [0]);
    }
    /**
     * @param username enka.network username, not in-game nickname
     * @param hash EnkaGameAccount hash
     * @returns the genshin account with provided hash
     */
    async fetchEnkaGenshinAccount(username, hash) {
        return await this.options.enkaSystem.fetchEnkaGameAccount(username, hash);
    }
    /**
     * @param username enka.network username, not in-game nickname
     * @param hash EnkaGameAccount hash
     * @returns the genshin character builds including saved builds in Enka.Network account
     */
    async fetchEnkaGenshinBuilds(username, hash) {
        return await this.options.enkaSystem.fetchEnkaCharacterBuilds(username, hash);
    }
    /**
     * @returns all playable character data
     */
    getAllCharacters() {
        return this.cachedAssetsManager.getGenshinCacheData("AvatarExcelConfigData").filterArray((_, p) => p.getAsStringWithDefault(null, "useType") === "AVATAR_FORMAL" && p.getAsNumber("featureTagGroupID") === p.getAsNumber("id")).map(([, p]) => characterUtils.getCharactersById(p.getAsNumber("id"), this)).reduce((a, b) => [...a, ...b], []);
    }
    /**
     * @param id The id of the character
     * @param skillDepotId Specifies one or zero elements for Traveler
     */
    getCharacterById(id, skillDepotId) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return CharacterData_1.CharacterData.getById(Number(id), this, skillDepotId ? Number(skillDepotId) : undefined);
    }
    /**
     * @returns all weapon data
     */
    getAllWeapons(excludeInvalidWeapons = true, filterByCodex = true) {
        const weapons = this.cachedAssetsManager.getGenshinCacheData("WeaponExcelConfigData");
        const weaponDataList = excludeInvalidWeapons
            ? weapons.filterArray((_, p) => p.has("id") && p.has("weaponPromoteId") && p.getAsNumber("weaponPromoteId") === p.getAsNumber("id")).map(([, p]) => new WeaponData_1.WeaponData(p.getAsJsonObject(), this))
            : weapons.mapArray((_, p) => new WeaponData_1.WeaponData(p.getAsJsonObject(), this));
        if (filterByCodex) {
            const codexSet = new Set(this.cachedAssetsManager.getGenshinCacheData("WeaponCodexExcelConfigData").mapArray((_, p) => p.getAsNumber("weaponId")));
            return weaponDataList.filter(w => codexSet.has(w.id));
        }
        else {
            return weaponDataList;
        }
    }
    /**
     * @param id The id of the weapon
     */
    getWeaponById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return WeaponData_1.WeaponData.getById(Number(id), this);
    }
    /**
     * @param includeDefaults Whether to include default costumes
     * @returns all costume data
     */
    getAllCostumes(includeDefaults = false) {
        return this.cachedAssetsManager.getGenshinCacheData("AvatarCostumeExcelConfigData").filterArray((_, p) => !includeDefaults || (includeDefaults && p.getAsBooleanWithDefault(false, "isDefault"))).map(([, p]) => new Costume_1.Costume(p.getAsJsonObject(), this));
    }
    /**
     * @param id The id of the costume
     */
    getCostumeById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return Costume_1.Costume.getById(Number(id), this);
    }
    /**
     * @returns all material data
     */
    getAllMaterials() {
        return this.cachedAssetsManager.getGenshinCacheData("MaterialExcelConfigData").mapArray((_, p) => Material_2.Material.getMaterialByData(p.getAsJsonObject(), this));
    }
    /**
     * @param id The id of the material
     */
    getMaterialById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return Material_2.Material.getMaterialById(Number(id), this);
    }
    /**
     * @returns all name card data
     */
    getAllNameCards() {
        return this.cachedAssetsManager.getGenshinCacheData("MaterialExcelConfigData").filterArray((_, p) => p.has("materialType") && p.getAsString("materialType") === Material_1.NameCard.MATERIAL_TYPE).map(([, p]) => new Material_1.NameCard(p.getAsJsonObject(), this));
    }
    /**
     * @param id The id of the name card
     */
    getNameCardById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        const material = Material_2.Material.getMaterialById(Number(id), this);
        if (material.materialType !== Material_1.NameCard.MATERIAL_TYPE)
            throw new Error(`Material ${material.id} is not a NameCard.`);
        return material;
    }
    /**
     * @param highestRarityOnly Whether to return only the rarest of artifacts of the same type
     * @returns all artifact data
     */
    getAllArtifacts(highestRarityOnly = false) {
        const validRarityMap = {};
        this.cachedAssetsManager.getGenshinCacheData("ReliquaryCodexExcelConfigData").forEachArray((_, c) => {
            const setId = c.getAsNumber("suitId");
            const stars = c.getAsNumber("level");
            if (highestRarityOnly) {
                if (validRarityMap[setId]) {
                    if (validRarityMap[setId][0] < stars)
                        validRarityMap[setId][0] = stars;
                }
                else {
                    validRarityMap[setId] = [stars];
                }
            }
            else {
                if (!(setId in validRarityMap))
                    validRarityMap[setId] = [];
                validRarityMap[setId].push(stars);
            }
        });
        const excludeSetIds = this.cachedAssetsManager.getGenshinCacheData("ReliquarySetExcelConfigData").filterArray((_, p) => p.getValue("disableFilter") === 1).map(([, p]) => p.getAsNumber("setId"));
        const artifacts = this.cachedAssetsManager.getGenshinCacheData("ReliquaryExcelConfigData").filterArray((_, p) => {
            if (!p.has("setId"))
                return false;
            const setId = p.getAsNumber("setId");
            if (excludeSetIds.includes(setId))
                return false;
            const stars = p.getAsNumber("rankLevel");
            return validRarityMap[setId].includes(stars);
        });
        const chunked = (0, config_file_js_1.separateByValue)(artifacts, ([, p]) => `${p.getAsNumber("setId")}-${p.getAsString("equipType")}-${p.getAsNumber("rankLevel")}`);
        return Object.values(chunked).map(chunk => new ArtifactData_1.ArtifactData(chunk[chunk.length - 1][1].getAsJsonObject(), this));
    }
    /**
     * @param id The id of the artifact
     */
    getArtifactById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return ArtifactData_1.ArtifactData.getById(Number(id), this);
    }
    /**
     * @returns all artifact set data
     */
    getAllArtifactSets() {
        const sets = this.cachedAssetsManager.getGenshinCacheData("ReliquarySetExcelConfigData").filterArray((_, p) => p.getValue("disableFilter") !== 1);
        return sets.map(([, p]) => new ArtifactSet_1.ArtifactSet(p.getAsJsonObject(), this));
    }
    /**
     * @param id The id of artifact set
     */
    getArtifactSetById(id) {
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        return ArtifactSet_1.ArtifactSet.getById(Number(id), this);
    }
    /**
     * Clear all running tasks in the client.
     */
    close() {
        this._tasks.forEach(task => clearTimeout(task));
    }
}
exports.EnkaClient = EnkaClient;
