// enka-network-api/dist/models/Element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Element = exports.elementList = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../errors/AssetsNotFoundError");
const TextAssets_1 = require("./assets/TextAssets");
exports.elementList = [null, "Fire", "Water", "Grass", "Electric", "Ice", null, "Wind", "Rock"];
class Element {
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsString("textMapId");
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("textMapContentTextMapHash"), enka);
    }
    static getByElementType(elementType, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("ManualTextMapConfigData").findArray((_, p) => p.getAsString("textMapId") === elementType)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Element", elementType);
        return new Element(json.getAsJsonObject(), enka);
    }
}
exports.Element = Element;
