// enka-network-api/dist/models/ProfilePicture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CharacterProfilePicture = exports.ProfilePicture = void 0;
const config_file_js_1 = require("config_file.js");
const ImageAssets_1 = require("./assets/ImageAssets");
const TextAssets_1 = require("./assets/TextAssets");
const Costume_1 = require("./character/Costume");
const AssetsNotFoundError_1 = require("../errors/AssetsNotFoundError");
class ProfilePicture {
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("iconPath"), enka);
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("nameTextMapHash"), enka);
        const keys = enka.cachedAssetsManager.getObjectKeysManager();
        this.type = json.getAsString(keys.profilePictureTypeKey);
    }
    static getById(id, enka) {
        var _a;
        const profilePicture = (_a = enka.cachedAssetsManager.getGenshinCacheData("ProfilePictureExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!profilePicture)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("ProfilePicture", id);
        const keys = enka.cachedAssetsManager.getObjectKeysManager();
        const type = profilePicture.getAsString(keys.profilePictureTypeKey);
        switch (type) {
            case "PROFILE_PICTURE_UNLOCK_BY_AVATAR":
            case "PROFILE_PICTURE_UNLOCK_BY_COSTUME":
                return new CharacterProfilePicture(profilePicture.getAsJsonObject(), enka);
            default:
                return new ProfilePicture(profilePicture.getAsJsonObject(), enka);
        }
    }
    /**
     * @deprecated
     */
    static getByOldFormat(characterId, costumeId, enka) {
        var _a;
        const iconType = costumeId === null ? "PROFILE_PICTURE_UNLOCK_BY_AVATAR" : "PROFILE_PICTURE_UNLOCK_BY_COSTUME";
        const referenceId = costumeId === null ? characterId : costumeId;
        const keys = enka.cachedAssetsManager.getObjectKeysManager();
        const profilePictureData = (_a = enka.cachedAssetsManager.getGenshinCacheData("ProfilePictureExcelConfigData").findArray((_, p) => p.getAsString(keys.profilePictureTypeKey) === iconType && p.getAsNumber("unlockParam") === referenceId)) === null || _a === void 0 ? void 0 : _a[1];
        return new CharacterProfilePicture(profilePictureData.getAsJsonObject(), enka);
    }
}
exports.ProfilePicture = ProfilePicture;
class CharacterProfilePicture extends ProfilePicture {
    constructor(data, enka) {
        super(data, enka);
        const json = new config_file_js_1.JsonReader(this._data);
        const keys = enka.cachedAssetsManager.getObjectKeysManager();
        const type = json.getAsString(keys.profilePictureTypeKey);
        if (type !== "PROFILE_PICTURE_UNLOCK_BY_AVATAR" && type !== "PROFILE_PICTURE_UNLOCK_BY_COSTUME")
            throw new Error("Invalid type for CharacterProfilePicture");
        const costume = type === "PROFILE_PICTURE_UNLOCK_BY_COSTUME"
            ? Costume_1.Costume.getById(json.getAsNumber("unlockParam"), enka)
            : Costume_1.Costume.getDefaultCostumeByCharacterId(json.getAsNumber("unlockParam"), enka);
        this.type = type;
        this.costume = costume;
    }
}
exports.CharacterProfilePicture = CharacterProfilePicture;
