// enka-network-api/dist/models/weapon/WeaponRefinement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaponRefinement = void 0;
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = require("../assets/TextAssets");
const StatProperty_1 = require("../StatProperty");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
class WeaponRefinement {
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.level = json.getAsNumberWithDefault(0, "level") + 1;
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("descTextMapHash"), enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.StatProperty(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.paramList = json.get("paramList").mapArray((_, p) => p.getAsNumber());
    }
    /**
     * @param level refinement rank (1-5)
     */
    static getById(id, level, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("EquipAffixExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id && p.getAsNumberWithDefault(0, "level") === level - 1)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("WeaponRefinement", id);
        return new WeaponRefinement(json.getAsJsonObject(), enka);
    }
}
exports.WeaponRefinement = WeaponRefinement;
