// enka-network-api/dist/models/character/talents/PassiveTalent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassiveTalent = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../../errors/AssetsNotFoundError");
const ImageAssets_1 = require("../../assets/ImageAssets");
const TextAssets_1 = require("../../assets/TextAssets");
const StatProperty_1 = require("../../StatProperty");
class PassiveTalent {
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("proudSkillId");
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("descTextMapHash"), enka);
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("icon"), enka);
        this.requiredAscension = json.getAsNumberWithDefault(0, "breakLevel");
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.StatProperty(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.isHidden = json.getAsBooleanWithDefault(false, "isHideLifeProudSkill");
    }
    static getById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("ProudSkillExcelConfigData").findArray((_, p) => p.getAsNumber("proudSkillId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Talent", id);
        return new PassiveTalent(json.getAsJsonObject(), enka);
    }
}
exports.PassiveTalent = PassiveTalent;
