// enka-network-api/dist/models/character/Constellation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constellation = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const ImageAssets_1 = require("../assets/ImageAssets");
const TextAssets_1 = require("../assets/TextAssets");
const StatProperty_1 = require("../StatProperty");
class Constellation {
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(data);
        this.id = json.getAsNumber("talentId");
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("descTextMapHash"), enka);
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("icon"), enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.StatProperty(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.paramList = json.get("paramList").mapArray((_, p) => p.getAsNumber());
    }
    static getById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("AvatarTalentExcelConfigData").findArray((_, p) => p.getAsNumber("talentId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Constellation", id);
        return new Constellation(json.getAsJsonObject(), enka);
    }
}
exports.Constellation = Constellation;
